### PR TITLE
🗃️ Check and copy missing local storage keys on start and resume

### PR DESCRIPTION
**Pull request by shankari**
This is the core of the workaround for
https://github.com/e-mission/e-mission-docs/issues/930

Essentially, we can store key-value data in both web-based local storage, and as a local storage type in the SQLite database. In practice, we have found that both of these are unreliable. On iOS, the webview based local storage is sometimes deleted, and on android, the SQLite reads sometimes return null.

Therefore, we plan to treat them as a RAID1 array. We will store identical copies of "local storage" data in both locations, and ensure that they are in sync on every app load and resume.

We keep them in sync by geting lists of keys for both storage locations, finding the missing ones in each and copying them over.

Testing done:

Bunch of keys are present:

```
unifiedlogger.js:49 DEBUG:STORAGE_PLUGIN: app has resumed, checking storage sync
storage.js:169 STORAGE_PLUGIN: Called syncAllWebAndNativeValues
storage.js:171 STORAGE_PLUGIN: native plugin returned
unifiedlogger.js:49 DEBUG:STORAGE_PLUGIN: Comparing web keys CONFIG_PHONE_UI,connection_settings with CONFIG_PHONE_UI,CONFIG_PHONE_UI,connection_settings
unifiedlogger.js:49 DEBUG:STORAGE_PLUGIN: Found native keys CONFIG_PHONE_UI,connection_settings missing native keys
unifiedlogger.js:49 DEBUG:STORAGE_PLUGIN: Found web keys CONFIG_PHONE_UI,CONFIG_PHONE_UI,connection_settings missing web keys
unifiedlogger.js:49 DEBUG:STORAGE_PLUGIN: Syncing all missing keys
unifiedlogger.js:49 DEBUG:STORAGE_PLUGIN: For the record, all unique native keys are stats/client_nav_event,stats/client_nav_event,background/battery,CONFIG_PHONE_UI,stats/client_time,config/app_ui_config,CONFIG_PHONE_UI,connection_settings,stats/client_time
startprefs.js:72 Not consented in local storage, need to show consent
2startprefs.js:76 Consented in local storage, no need to show consent
```

From the profile screen, use the `Nuke all buffers and cache` option to delete all usercache information.
On resume, there are no native keys, but there are a bunch of web keys
We copy all of them over.

```
unifiedlogger.js:49 DEBUG:STORAGE_PLUGIN: app has resumed, checking storage sync
storage.js:169 STORAGE_PLUGIN: Called syncAllWebAndNativeValues
storage.js:171 STORAGE_PLUGIN: native plugin returned
unifiedlogger.js:49 DEBUG:STORAGE_PLUGIN: Comparing web keys intro_done,connection_settings,data_collection_consented_protocol,CONFIG_PHONE_UI,prompted-auth with
unifiedlogger.js:49 DEBUG:STORAGE_PLUGIN: Found native keys  missing native keys intro_done,connection_settings,data_collection_consented_protocol,CONFIG_PHONE_UI,prompted-auth
unifiedlogger.js:49 DEBUG:STORAGE_PLUGIN: Found web keys  missing web keys
unifiedlogger.js:49 DEBUG:STORAGE_PLUGIN: Syncing all missing keys intro_done,connection_settings,data_collection_consented_protocol,CONFIG_PHONE_UI,prompted-auth
unifiedlogger.js:49 DEBUG:STORAGE_PLUGIN: For the record, all unique native keys are
```

On the next resume, they are all there

```
DEBUG:STORAGE_PLUGIN: app has resumed, checking storage sync
storage.js:169 STORAGE_PLUGIN: Called syncAllWebAndNativeValues
storage.js:171 STORAGE_PLUGIN: native plugin returned
unifiedlogger.js:49 DEBUG:STORAGE_PLUGIN: Comparing web keys intro_done,connection_settings,CURR_GEOFENCE_LOCATION,data_collection_consented_protocol,CONFIG_PHONE_UI,prompted-auth with intro_done,connection_settings,CURR_GEOFENCE_LOCATION,data_collection_consented_protocol,CONFIG_PHONE_UI,prompted-auth
unifiedlogger.js:49 DEBUG:STORAGE_PLUGIN: Found native keys intro_done,connection_settings,CURR_GEOFENCE_LOCATION,data_collection_consented_protocol,CONFIG_PHONE_UI,prompted-auth missing native keys
unifiedlogger.js:49 DEBUG:STORAGE_PLUGIN: Found web keys intro_done,connection_settings,CURR_GEOFENCE_LOCATION,data_collection_consented_protocol,CONFIG_PHONE_UI,prompted-auth missing web keys
unifiedlogger.js:49 DEBUG:STORAGE_PLUGIN: Syncing all missing keys
unifiedlogger.js:49 DEBUG:STORAGE_PLUGIN: For the record, all unique native keys are intro_done,connection_settings,CURR_GEOFENCE_LOCATION,data_collection_consented_protocol,
CONFIG_PHONE_UI,prompted-auth,stats/client_time
```

### CODE REVIEW

**Comment by shankari** [comment]
Steps for testing:
- Launch app
- Login using a stage-study URL
- Go through the onboarding
- Go to the profile
- "Nuke all buffers and cache"
- Move app to background and then foreground
- Get several alerts about native value missing and local value found
- Go to the label screen, resync, trips load again

| Screen 1 | Screen 2 | Screen 3 |
|-----|------|----|
| Screenshot 2023-07-11 at 9 04 16 PM | Screenshot 2023-07-11 at 9 04 10 PM | Screenshot 2023-07-11 at 9 04 04 PM |
| Screenshot 2023-07-11 at 9 03 58 PM | Screenshot 2023-07-11 at 9 03 52 PM |

**Comment by shankari** [comment]
Final issue to investigate before I merge this:
- Although all the testing steps work correctly, if I go back to the profile screen after all this, the app logs out and returns to the initial screen.
- It stays on the initial screen even if killed in relaunched
- However, the logs seem to indicate that the config **is** present as `CONFIG_PHONE_UI`
- Need to investigate further...


```
DEBUG:STORAGE_PLUGIN: app launched, checking storage sync
storage.js:169 STORAGE_PLUGIN: Called syncAllWebAndNativeValues 
storage.js:171 STORAGE_PLUGIN: native plugin returned
unifiedlogger.js:49 DEBUG:STORAGE_PLUGIN: Comparing web keys intro_done,connection_settings,CURR_GEOFENCE_LOCATION,data_collection_consented_protocol,CONFIG_PHONE_UI,prompted-auth with intro_done,connection_settings,CURR_GEOFENCE_LOCATION,data_collection_consented_protocol,CONFIG_PHONE_UI,prompted-auth,CURR_GEOFENCE_LOCATION,CURR_GEOFENCE_LOCATION
unifiedlogger.js:49 DEBUG:STORAGE_PLUGIN: Found native keys intro_done,connection_settings,CURR_GEOFENCE_LOCATION,data_collection_consented_protocol,CONFIG_PHONE_UI,prompted-auth missing native keys 
unifiedlogger.js:49 DEBUG:STORAGE_PLUGIN: Found web keys intro_done,connection_settings,CURR_GEOFENCE_LOCATION,data_collection_consented_protocol,CONFIG_PHONE_UI,prompted-auth,CURR_GEOFENCE_LOCATION,CURR_GEOFENCE_LOCATION missing web keys 
unifiedlogger.js:49 DEBUG:STORAGE_PLUGIN: Syncing all missing keys 
unifiedlogger.js:49 DEBUG:STORAGE_PLUGIN: For the record, all unique native keys are intro_done,connection_settings,CURR_GEOFENCE_LOCATION,data_collection_consented_protocol,CONFIG_PHONE_UI,prompted-auth,stats/client_time,stats/client_time,config/consent,stats/client_time,statemachine/transition,background/battery,CURR_GEOFENCE_LOCATION,statemachine/transition,statemachine/transition,background/battery,CURR_GEOFENCE_LOCATION,stats/client_nav_event
3startprefs.js:76 Consented in local storage, no need to show consent
```

**Comment by shankari** [comment]
On launch, I do see the logs `for key CONFIG_PHONE_UI uc_stored_val = {"version":1,"ts":1655143472,"server":{"connectUrl":"https://openpath-stage.nrel.gov/api/","aggregate_call_auth":"user_only"},` but

```
DEBUG:in readConfig, read_val = {}
unifiedlogger.js:49 DEBUG:Config not downloaded, need to show join screen
```

So we need to focus on readConfig.

Aha! It is because we use the config value in multiple places

```
www/js//splash/startprefs.js|20| var CONFIGURED_KEY = "config/app_ui_config";
www/js//config/dynamic_config.js|11| const CONFIG_PHONE_UI="config/app_ui_config";
www/js//intro.js|85| const CONFIG_PHONE_UI="config/app_ui_config";
www/js//control/general-settings.js|411| const CONFIG_PHONE_UI="config/app_ui_config";
```

Need to fix all of them, not just the dynamic config.

**Comment by shankari** [comment]
@JGreenlee would you like to review this? Note the other native code changes linked to https://github.com/e-mission/e-mission-docs/issues/930 as well. I will plan on merging and deploying today afternoon.

File: www/js/control/general-settings.js [hunk 1]
```diff
@@ -408,9 +408,7 @@ angular.module('emission.main.control',['emission.services',
             if (!res) return; // user cancelled
             
             // reset the saved config, then trigger a hard refresh
-            const CONFIG_PHONE_UI="config/app_ui_config";
-            $window.cordova.plugins.BEMUserCache.putRWDocument(CONFIG_PHONE_UI, {})
-                .then($window.location.reload(true));
+            DynamicConfig.resetConfigAndRefresh();
```
Comment (JGreenlee) [inline www/js/control/general-settings.js:411]: It was a good idea to extract this to a separate function anyway

**Review by JGreenlee — APPROVED** [review]
Looks good, and I think I understand at a high level.
I'd imagine this is relatively temporary - hopefully after the migration we won't have to store data in both places and sync them, right?

**Comment by JGreenlee** [comment]
The native code is pretty foreign to me, but I did at least give it a cursory look-over.

It seemed like the main change is here; the other changes were for removing `email`- related code since we only use OPCodes now, and adding a few methods to the usercache plugin for debugging purposes

**Comment by shankari** [comment]
The methods to the usercache plugin are not just for debugging, they are actually used here.
https://github.com/e-mission/e-mission-phone/pull/1003/files#diff-7063ecbf9a71714bf3f2348875db819f212bc6b6fe808a17e4fc850968d84735R170

We need to read the full list of keys stored in SQLite so that we can see which are missing and need to be copied over.

**Comment by shankari** [comment]
>  I'd imagine this is relatively temporary - hopefully after the migration we won't have to store data in both places and sync them, right?

I am not sure that this is temporary. After the migration, we might switch to
https://github.com/react-native-async-storage/async-storage
instead of the webview-based local storage and see how stable that is.

If it is stable, I am happy to keep remove the duplication. But if (like the webview-based local storage), it can be arbitrarily deleted by the OS to free up space, I would rather keep the duplication. As you can see from the issue, we rely a lot on this storage to track the opcode, which study people are logged into, etc. and having that data be lost is completely catastrophic.

**Comment by shankari** [comment]
> I am not sure what happens to other apps when their data is lost - maybe users are just logged out. Maybe we can message it accordingly in our app as well. If we do that, we can remove the duplication code.

For the record, one of the challenges with "users are just logged out" is that it works well for services that users interact with a lot, and not as well for background services. If we are logged out; the data won't be pushed to the server if/until the user logs in again.

**Pull request merged**